docs/proposals/odo-deploy.md

```diff
@@ -0,0 +1,148 @@
+1. Edit the project source code to develop the application.
+1. `odo url create` - This stores URL information (host, port etc.) for accessing the application on the cluster (if not done already).
+1. `odo push` - This runs the application source code using inner-loop instructions from devfile.
+1. Validate the running application is working as intended.
+1. Iterate over steps 2 and beyond (as needed). 
+1. `odo deploy` - This would build a new clean image and deploy it to the target cluster using outer-loop instructions from devfile and user-provided arguments.
+1. Validate the deployed application is working as intended.
+1. Iterate over steps 2 and beyond (as needed).
+1. Optionally, run `odo deploy delete` to clean up resources created with `odo deploy`.
+1. Push your code to Git - ready for sharing with your team and for CI/CD pipelines to take over.
+
+## User stories
+
+### Initial build and deploy support to odo - https://github.com/openshift/odo/issues/3300
+
+## Design overview
+`odo deploy` could provide developers with a way to build a container image for their application and deploy it on a target Kubernetes deployment using the build/deploy guidance provided by the devfile.
+
+This deployment is equivalent to a development version of your production and will be using the namespace and URL information from the inner-loop. This will ensure that it is not seen as a way to deploy real workloads in production.
+
+`odo deploy delete` could provide developers with a way to clean up any existing deployment of the application. 
+
+### High-level design:
+
+#### Pre-requisites:
+- The implementation would be under the experimental flag.
+- Only supported for Devfile v2.0.0 components.
+- Only supported for Kubernetes/OpenShift targets.
+
+#### odo deploy 
+This command will build a container image for the application and deploy it on the target Kubernetes environment. 
+
+Flags:
+ - `--tag`: The tag to be used for the built application container image - `<registry>/<org>/<name>:<tag>` (optional)
+    - Openshift: the tag can be generated based on project details and internal image registry can be used
+    - Kubernetes: the user must provide the fully qualified registry and tag details.
+ - `--credentials`: The credentials needed to push the image to the container image registry (optional)
+    - Openshift: default builder account for BuildConfig is used
+    - Kubernetes: the user must provide the credentials
+ - `--force`: Delete the existing deployment (if present) and re-create a new deployment.
+
+#### odo deploy delete
+This command will delete any resources created by odo deploy.
+
+### Detailed design: 
+
+### Devfile
+
+For the initial implementation, we could use devfile v2.0.0 and capture basic outer-loop information as `metadata` on the devfile. `odo` could look for  specific keys, while other tools like Che could ignore them.
+
+For example: 
+```
+metadata:
+    build-dockerfile: <URI>
```

Comment:
Would it make sense to highlight the fact that this is probably just a temporary location that will be replaced with proper fields in future devfile spec versions?
Kubernetes API does something similar where they use annotations containing `alpha` group before they settle down on the proper placement.

For example, we could do something like this:
```yaml
metadata:
  alpha.build-dockerfile
  alpha.deployment-manifest
```

Reply:
Makes sense. I'll update the proposal to reflect this.